src/core/geometry/qgspolyhedralsurface.cpp

```diff
@@ -0,0 +1,997 @@
+{
+  // sum area of patches
+  double area = 0.0;
+  for ( const QgsPolygon *patch : mPatches )
+  {
+    area += patch->area();
+  }
+
+  return area;
+}
+
+double QgsPolyhedralSurface::perimeter() const
+{
+  // sum perimeter of patches
+  double perimeter = 0.0;
+  for ( const QgsPolygon *patch : mPatches )
+  {
+    perimeter += patch->perimeter();
+  }
+
+  return perimeter;
+}
+
+QgsAbstractGeometry *QgsPolyhedralSurface::boundary() const
+{
+  std::unique_ptr< QgsMultiLineString > multiLine( new QgsMultiLineString() );
+  multiLine->reserve( mPatches.size() );
+  for ( QgsPolygon *polygon : mPatches )
+  {
+    QgsAbstractGeometry *polygonBoundary = polygon->boundary();
+    if ( QgsLineString *lineStringBoundary = qgsgeometry_cast< QgsLineString * >( polygonBoundary ) )
+    {
+      multiLine->addGeometry( lineStringBoundary );
+    }
+    else if ( QgsMultiLineString *multiLineStringBoundary = qgsgeometry_cast< QgsMultiLineString * >( polygonBoundary ) )
+    {
+      for ( int j = 0; j < multiLineStringBoundary->numGeometries(); ++j )
+      {
+        multiLine->addGeometry( multiLineStringBoundary->geometryN( j )->clone() );
+      }
+      delete multiLineStringBoundary;
+    }
+    else
+    {
+      delete polygonBoundary;
+    }
+  }
+
+  if ( multiLine->numGeometries() == 0 )
+  {
+    return nullptr;
+  }
+  return multiLine.release();
+}
+
+QgsPolyhedralSurface *QgsPolyhedralSurface::snappedToGrid( double hSpacing, double vSpacing, double dSpacing, double mSpacing, bool removeRedundantPoints ) const
+{
+  std::unique_ptr< QgsPolyhedralSurface > surface( createEmptyWithSameType() );
+
+  for ( QgsPolygon *patch : mPatches )
+  {
+    // exterior ring
+    QgsCurve *exteriorRing = static_cast< QgsCurve *>( patch->exteriorRing()->snappedToGrid( hSpacing, vSpacing, dSpacing, mSpacing, removeRedundantPoints ) );
+    if ( !exteriorRing )
+    {
+      return nullptr;
+    }
+
+    QgsPolygon *gridifiedPatch = new QgsPolygon;
+    gridifiedPatch->setExteriorRing( exteriorRing );
+
+    //interior rings
+    for ( int i = 0, n = patch->numInteriorRings(); i < n; ++i )
+    {
+      QgsCurve *interiorRing = patch->interiorRing( i );
+      if ( !interiorRing )
+        continue;
+
+      QgsCurve *gridifiedInterior = static_cast< QgsCurve * >( interiorRing->snappedToGrid( hSpacing, vSpacing, dSpacing, mSpacing, removeRedundantPoints ) );
+      if ( !gridifiedInterior )
+        continue;
+
+      gridifiedPatch->addInteriorRing( gridifiedInterior );
+    }
+
+    surface->addPatch( gridifiedPatch );
+  }
+
+  return surface.release();
+}
+
+QgsPolyhedralSurface *QgsPolyhedralSurface::simplifyByDistance( double tolerance ) const
+{
+  if ( isEmpty() )
+    return nullptr;
+
+  std::unique_ptr< QgsPolyhedralSurface > simplifiedGeom = std::make_unique< QgsPolyhedralSurface >();
+  for ( QgsPolygon *polygon : mPatches )
+  {
+    simplifiedGeom->addPatch( polygon->simplifyByDistance( tolerance )->surfaceToPolygon() );
```

Comment:
leaks the geometry returned by simplifyByDistance

Reply:
Fixed.